Parser/tokenizer.c

```diff
@@ -2072,6 +2072,79 @@ _PyTokenizer_Get(struct tok_state *tok,
     return result;
 }
 
+#if defined(__wasi__) || defined(__EMSCRIPTEN__)
+/* fdopen() with borrowed fd
+
+   WASI does not provide dup() and Emscripten's dup() emulation with open()
+   is slow. Implement fdopen() with fd borrowing on top of fdopencookie().
+ */
+typedef union {
+    void *cookie;
+    int fd;
+} borrowed;
+
+static ssize_t
+borrow_read(void *cookie, char *buf, size_t size)
+{
+    borrowed b;
+    b.cookie = cookie;
+    return read(b.fd, (void *)buf, size);
+}
+
+static ssize_t
+borrow_write(void *cookie, const char *buf, size_t size)
+{
+    errno = ENOTSUP;
+    return -1;
+}
+
+static int
+borrow_seek(void *cookie, off_t *off, int whence)
+{
+    borrowed b;
+    b.cookie = cookie;
+    off_t pos;
+    pos = lseek(b.fd, *off, whence);
+    if (pos == (off_t)-1) {
+        return -1;
+    } else {
+        *off = pos;
+        return 0;
+    }
+}
+
+static int
+borrow_close(void *cookie)
+{
+    // does not close(fd)
+    return 0;
+}
+
+static FILE *
+fdopen_borrow(int fd, const char *mode) {
+    // only reading is supported
+    if (strcmp(mode, "r") != 0) {
+        return NULL;
+    }
+    cookie_io_functions_t cookie_io = {
+        borrow_read, borrow_write, borrow_seek, borrow_close
+    };
+    // cookie is just the fd
+    borrowed b;
+    b.fd = fd;
+    return fopencookie(b.cookie, "r", cookie_io);
```

Comment:
`fopencookie` is a non-standard GNU extension. Is this available in all platforms targetted by web assembly? Does this need a configure check?

Reply:
I had a configure check in my first version of the patch. I got rid of it because it makes no sense. Both WASI and Emscripten use musl libc as upper half of libc and it always provides fdopencookie.

I wish their was a more elegant way to solve the problem. ``_PyTokenizer_FindEncodingFilename`` uses a FILE pointer instead of a file descriptor.

Reply:
Ugh, I am honestly not very happy with the amount of extra code, but I understand why is needed.